auth/service/api/v1/devicetokens.go

```diff
@@ -0,0 +1,96 @@
+package v1
+
+import (
+	"fmt"
+	"log"
+	"net/http"
+
+	"github.com/ant0ine/go-json-rest/rest"
+
+	"github.com/tidepool-org/platform/devicetokens"
+	"github.com/tidepool-org/platform/request"
+	"github.com/tidepool-org/platform/service/api"
+)
+
+func (r *Router) DeviceTokensRoutes() []*rest.Route {
+	return []*rest.Route{
+		rest.Post("/v1/users/:userId/device_tokens", api.RequireAuth(r.UpsertDeviceToken)),
+	}
+}
+
+func (r *Router) UpsertDeviceToken(res rest.ResponseWriter, req *rest.Request) {
+	responder := request.MustNewResponder(res, req)
+	ctx := req.Request.Context()
+	authDetails := request.GetAuthDetails(ctx)
+	repo := r.AuthStore().NewDeviceTokenRepository()
+
+	if err := checkAuthentication(authDetails); err != nil {
+		log.Printf("checkAuth failed: %+v", authDetails)
+		responder.Error(http.StatusUnauthorized, err)
+		return
+	}
+
+	if err := checkUserIDConsistency(authDetails, req.PathParam("userId")); err != nil {
+		log.Printf("checkUserIDConsistency failed: %+v %q", authDetails, req.PathParam("userID"))
+		responder.Error(http.StatusForbidden, err)
+		return
+	}
+
+	deviceToken := devicetokens.DeviceToken{}
+	if err := request.DecodeRequestBody(req.Request, &deviceToken); err != nil {
+		responder.Error(http.StatusBadRequest, err)
+		return
+	}
+
+	userID := userIDWithServiceFallback(authDetails, req.PathParam("userId"))
+	doc := devicetokens.NewDocument(userID, deviceToken)
+	if err := repo.Upsert(ctx, doc); err != nil {
+		responder.Error(http.StatusInternalServerError, err)
+		return
+	}
+}
+
+var ErrUnauthorized = fmt.Errorf("unauthorized")
```

Comment:
`fmt.Errorf` is fine, so long as it wraps a platform error though, right?

Reply:
No, unfortunately not. Using `fmt.Errorf` will not include the full context of the error. When you use the platform `errors.Wrap` (or `errors.Wrapf`) it is maintaining a hierarchy of errors so that when it is logged that hierarchy is visible. It isn't particularly useful if at the lowest level, say Mongo, it has a "unable to get data" without knowing the "error stack" above that. This is one reason why it is convention to wrap the error at each step as the error bubbles up.

Remember that platform errors existed *well* before the Golang errors package had the ability to maintain an error hierarchy (using `%w`, `Unwrap`, etc.). The initial version of platform uses Go 1.6 while the Golang errors hierarchy was added in Go 1.13.

Should this be updated? Most definitely. Has it been? No, not yet. At some point we should reconcile the differences between the two errors packages, but we haven't yet.

Reply:
> No, unfortunately not. Using fmt.Errorf will not include the full context of the error.

Isn't that what `errors.As` is for? https://pkg.go.dev/errors#As

Granted that wasn't in Go 1.6, a point which I totally understand, but that should be an easy fix, to add a errors.As to get the wrapped platform error with it's full context?

Reply:
I'm not sure I understand. How would we use `errors.As` here?

Reply:
But yeah, I understand that doesn't exist right now, so no fmt.Errorf until it is... if it ever is.

Reply:
> I'm not sure I understand. How would we use errors.As here?

As I understand it, it would work something like this:

```go
func foo() {
	myErr := fmt.Errorf("I added some context information: %w", request.ErrBadRequest())
	// Then later on, we want to retrieve the context from
	// request.ErrBadRequest...  This assumes that we're within the errors
	// package for now, as object isn't exported.
	reqErr := &object{}
	if errors.As(myErr, reqErr) {
		// errors.As returned true, which means that it was able to unwrap
		// errors around myErr and found an object{} in there somewhere. I can
		// now access that error via reqErr.
		fmt.Println("contrived example", object.Detail, object.Caller)		
	}
}
```

Though in writing that, I just saw that there's an `errors.Wrapf`, and that would serve the same purpose as fmt.Errorf in this case, but using the machinations of the existing errors package, so... yeah.